### PR TITLE
Clean up usage of string.IndexOf / ToUpper / ToLower / Trim throughout the framework

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
__High-level summary of changes:__

 * Improve performance of `string.ToUpperInvariant`, `string.ToLowerInvariant`, `char.ToUpperInvariant`, `char.ToLowerInvariant` by binding them directly to the correct `TextInfo` instance rather than bouncing through `CultureInfo.Invariant`.

 * Avoid dereferencing `CultureInfo.InvariantCulture` where possible at the call sites and use the invariant-specific code paths.

 * Replace the pattern `if (string.Trim().Length != 0)` to use `string.IsNullOrWhiteSpace` instead.

 * Replace call sites that incorrectly use culture-aware `string.ToUpper` / `string.ToLower` to use `ToUpperInvariant` / `ToLowerInvariant` instead.

 * Replace call sites that incorrectly use culture-aware `string.IndexOf(...)` to use `string.IndexOf(..., StringComparison.Ordinal)` instead.

 * Replace call sites that incorrectly use culture-aware `string.StartsWith(...)` / `string.EndsWith(...)` to use ordinal `string.StartsWith(..., StringComparison.Ordinal)` / `string.EndsWith(..., StringComparison.Ordinal)` instead.

 * If a project specifically targets .NET Core latest, change `string.IndexOf("x", StringComparison.Ordinal)` to `string.IndexOf('x')`.

There's also lots of opportunity for performance improvements here, such as avoiding allocations of temporary objects. I wasn't really focusing on that as much as I was the low-hanging fruit of making sure the correct comparison enum value was passed into the call sites.

One thing that made this a bit difficult is that `string` and `ROS<char>` behave differently as far as culture-awareness of certain methods. For example:

```cs
string s = GetString();
int idxA = s.IndexOf("target"); // <-- CurrentCulture

ReadOnlySpan<char> span = s;
int idxB = span.IndexOf("target"); // <-- Ordinal
```

So if you don't know whether the _this_ parameter provided to these methods is a `string` or a `ReadOnlySpan<char>`, you won't know which `StringComparison` is being used.

### CODE REVIEW

File: src/libraries/Microsoft.XmlSerializer.Generator/src/Sgen.cs [hunk 1]
```diff
@@ -309,7 +309,7 @@ private void GenerateFile(List<string> typeNames, string assemblyName, bool prox
                     //However when a path ending with backslash, if followed by double quote, it becomes an escapte sequence
                     //e.g. "obj\Debug\netcoreapp2.0\", it will be converted as obj\Debug\netcoreapp2.0", which is not valid and not exist
                     //We need remove the ending quote for this situation
-                    if (!outputDirectory.EndsWith("\"") || !Directory.Exists(outputDirectory = outputDirectory.Remove(outputDirectory.Length - 1)))
+                    if (!outputDirectory.EndsWith("\"", StringComparison.Ordinal) || !Directory.Exists(outputDirectory = outputDirectory.Remove(outputDirectory.Length - 1)))
```
Comment (stephentoub) [inline src/libraries/Microsoft.XmlSerializer.Generator/src/Sgen.cs]: Could this just use `EndsWith('\')`?

File: src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/IdnElement.cs [hunk 2]
```diff
@@ -50,11 +50,9 @@ public override bool CanConvertFrom(ITypeDescriptorContext context, Type sourceT
 
             public override object ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, object value)
             {
-                string s = value as string;
-                if (s != null)
+                if (value is string s)
                 {
-                    s = s.ToLower(CultureInfo.InvariantCulture);
-                    switch (s)
+                    switch (s.ToLowerInvariant())
```
Comment (stephentoub) [inline src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/IdnElement.cs]: Is this allocation better than three `Equals(..., StringComparison.InvariantCultureIgnoreCase)` calls?

File: src/libraries/Microsoft.XmlSerializer.Generator/src/Sgen.cs (hunk 1 above)
Comment (GrabYourPitchforks) [reply src/libraries/Microsoft.XmlSerializer.Generator/src/Sgen.cs]: Doesn't exist in netstandard2.0, which this project targets. :(

File: src/libraries/System.Diagnostics.TextWriterTraceListener/src/System/Diagnostics/DelimitedListTraceListener.cs [hunk 3]
```diff
@@ -243,7 +243,7 @@ private void EscapeMessage(string message, StringBuilder sb)
         {
             int index;
             int lastindex = 0;
-            while ((index = message.IndexOf('"', lastindex)) != -1)
+            while ((index = message.IndexOf('\"', lastindex)) != -1)
```
Comment (stephentoub) [inline src/libraries/System.Diagnostics.TextWriterTraceListener/src/System/Diagnostics/DelimitedListTraceListener.cs]: I'm not clear what's happening here.  Why is this change necessary?  Does this even compile?

File: src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Utils.cs [hunk 4]
```diff
@@ -1935,10 +1935,10 @@ internal static string SplitServerNameAndPortNumber(string serverName, out strin
             }
 
             //extract IPv6 port number if any
-            bool isBrace = serverName.StartsWith("[");
+            bool isBrace = serverName.StartsWith("[", StringComparison.Ordinal);
```
Comment (stephentoub) [inline src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Utils.cs]: Is this netstandard2.0 and that's why we can't just pass '['?

File: src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/IdnElement.cs (hunk 2 above)
Comment (GrabYourPitchforks) [reply src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/IdnElement.cs]: This would have been a behavioral change. Had they used `ToUpperInvariant` instead of `ToLowerInvariant`, then using `Equals(..., OrdinalIgnoreCase)` would behave the same.

File: src/libraries/System.IO.Packaging/src/System/IO/Packaging/ContentType.cs [hunk 5]
```diff
@@ -403,7 +403,7 @@ private static int GetLengthOfParameterValue(string s, int startIndex)
 
                 while (!found)
                 {
-                    length = s.IndexOf('"', ++length);
+                    length = s.IndexOf('\"', ++length);
```
Comment (stephentoub) [inline src/libraries/System.IO.Packaging/src/System/IO/Packaging/ContentType.cs]: Why is this necessary?

File: src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Utils.cs (hunk 4 above)
Comment (GrabYourPitchforks) [reply src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Utils.cs]: Yeah, unfortunately. :(
https://github.com/dotnet/runtime/blob/fcfb4b8a62b0bc30d9ed83a08cf9b322908c3973/src/libraries/System.DirectoryServices/src/System.DirectoryServices.csproj#L7

File: src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/IdnElement.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/IdnElement.cs]: Why is it a behavioral change?

File: src/libraries/System.Net.HttpListener/src/System/Net/HttpListenerRequest.cs [hunk 6]
```diff
@@ -343,7 +343,7 @@ internal static string GetCharSetValueFromHeader(string headerValue)
                 {
                     if (i == l - 1)
                         return null;
-                    j = headerValue.IndexOf('"', i + 1);
+                    j = headerValue.IndexOf('\"', i + 1);
```
Comment (stephentoub) [inline src/libraries/System.Net.HttpListener/src/System/Net/HttpListenerRequest.cs]: Same question as some others... why is this necessary?

File: src/libraries/System.Private.CoreLib/src/System/Char.cs [hunk 7]
```diff
@@ -384,7 +384,7 @@ public static char ToLower(char c)
         // Converts a character to lower-case for invariant culture.
         public static char ToLowerInvariant(char c)
         {
-            return CultureInfo.InvariantCulture.TextInfo.ToLower(c);
+            return TextInfo.Invariant.ToLower(c);
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Char.cs]: I'd assumed I guess incorrectly that devirtualization was kicking in and enabling this to be direct calls... but I guess not?

File: src/libraries/System.Private.CoreLib/src/System/Environment.Unix.cs [hunk 8]
```diff
@@ -255,7 +255,7 @@ private static string ReadXdgDirectory(string homeDir, string key, string fallba
                             }
 
                             // Find end of path
-                            int endPos = line.IndexOf('"', pos);
+                            int endPos = line.IndexOf('\"', pos);
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Environment.Unix.cs]: Why?

**Comment by GrabYourPitchforks** [comment]
@terrajobst Steve said you were looking at new analyzer rules we should enable. Many of the things I changed here would have been caught by https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1307, but that rule is very noisy because it uses overly broad pattern-matching and has a high false positive rate. But there's possibly a need for a rule that triggers on just the APIs we know to be problematic, like the specific signatures `string.IndexOf(string)` and `string.StartsWith(string)`.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/TextInfo.cs [hunk 9]
```diff
@@ -45,9 +45,7 @@ private enum Tristate : byte
         private Tristate _isAsciiCasingSameAsInvariant = Tristate.NotInitialized;
 
         // Invariant text info
-        internal static TextInfo Invariant => s_invariant ??= new TextInfo(CultureData.Invariant);
-
-        private static volatile TextInfo? s_invariant;
+        internal static readonly TextInfo Invariant = new TextInfo(CultureData.Invariant);
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/TextInfo.cs]: Does this have any negative impact on startup?

File: src/libraries/System.Private.Uri/src/System/UriScheme.cs [hunk 10]
```diff
@@ -159,7 +159,7 @@ public static void Register(UriParser uriParser, string schemeName, int defaultP
             if ((defaultPort >= 0xFFFF || defaultPort < 0) && defaultPort != -1)
                 throw new ArgumentOutOfRangeException(nameof(defaultPort));
 
-            schemeName = schemeName.ToLower();
+            schemeName = schemeName.ToLowerInvariant();
```
Comment (stephentoub) [inline src/libraries/System.Private.Uri/src/System/UriScheme.cs]: This doesn't need to be culture-aware?

**Review by stephentoub — APPROVED** [review]
Thanks.

File: src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/IdnElement.cs (hunk 2 above)
Comment (GrabYourPitchforks) [reply src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/IdnElement.cs]: It's described in a little more detail at https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1308, but the short of it is that there are some inputs for which `a.ToUpperInvariant() == b.ToUpperInvariant()` but `a.ToLowerInvariant() != b.ToLowerInvariant()`. On Windows, an `OrdinalIgnoreCase` comparison normalizes to uppercase before performing the comparison. Linux today does the same thing, but I opened the issue https://github.com/dotnet/runtime/issues/26961 to reconsider this behavior since the Unicode standard does not recommend doing this.

File: src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Utils.cs (hunk 4 above)
Comment (danmosemsft) [reply src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Utils.cs]: Why is it netstandard at all?

File: src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/IdnElement.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/IdnElement.cs]: > some inputs

What about for these three specific strings, though? There are inputs that would return true when compared with these three strings when ignoring case invariantly but that would return false when lowercased and then compared? What's an example?

Comment (GrabYourPitchforks) [reply src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/IdnElement.cs]: For these three specific strings, at this specific moment in time, nothing. However:

 * If https://github.com/dotnet/runtime/pull/27540 goes through as-is, `'İ'` will lowercase to `'i'` when running on Linux (ICU).
 * Certain other non-ASCII characters already lowercase to ASCII characters. For instance, under ICU today, _three_ total characters (`'k'`, `'K'`, and `'K' (U+212A KELVIN SIGN)`) all lowercase to `'k'`. However, only _two_ characters (`'k'` and `'K'`) uppercase to `'K'`. I realize this isn't one of the characters in these strings, but it does demonstrate that "normalize to uppercase" and "normalize to lowercase" are behaviorally inconsistent.
 * There's nothing stopping Unicode from tomorrow introducing a new character that lowercases to one of the ASCII characters contained within these three strings.

The risk is probably low enough where it's not worth worrying about. But with this PR I was trying to avoid introducing behavioral changes beyond "this uses a culture-aware conversion but it really should have been non-linguistic instead."

File: src/libraries/System.Diagnostics.TextWriterTraceListener/src/System/Diagnostics/DelimitedListTraceListener.cs (hunk 3 above)
Comment (GrabYourPitchforks) [reply src/libraries/System.Diagnostics.TextWriterTraceListener/src/System/Diagnostics/DelimitedListTraceListener.cs]: This (and others like it) was just general cleanup. I can revert if desired.

Comment (stephentoub) [reply src/libraries/System.Diagnostics.TextWriterTraceListener/src/System/Diagnostics/DelimitedListTraceListener.cs]: Personally I'd prefer if it was reverted; the original seems better to me.

File: src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/IdnElement.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Configuration.ConfigurationManager/src/System/Configuration/IdnElement.cs]: Ok.  It's fine, thanks.

File: src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Utils.cs (hunk 4 above)
Comment (terrajobst) [reply src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Utils.cs]: Because it's part of the Windows Compat Pack, which shipped for .NET Standard 2.0/.NET Standard 2.0. We decided that we'd rather have a turd assembly then having to deal with disjoint API surface area between Windows and non-Windows (i.e. a throwing is API is better than an API that is missing based on operating system).

I don't think we can change this for .NET 5 because customers building for .NET Standard 2.0 are expected to not lose any APIs when retargeting to `net5`.

File: src/libraries/System.Diagnostics.TextWriterTraceListener/src/System/Diagnostics/DelimitedListTraceListener.cs (hunk 3 above)
Comment (GrabYourPitchforks) [reply src/libraries/System.Diagnostics.TextWriterTraceListener/src/System/Diagnostics/DelimitedListTraceListener.cs]: No problem. :)

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs [hunk 11]
```diff
@@ -586,15 +586,15 @@ public static ulong GetKey(ReadOnlySpan<byte> propertyName)
 
             if (type.FullName != null)
             {
-                if (type.FullName.StartsWith("System.Collections.Generic.IEnumerable`1"))
+                if (type.FullName.StartsWith("System.Collections.Generic.IEnumerable`1", StringComparison.Ordinal))
```
Comment (steveharter) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs]: cc @layomia

File: src/libraries/System.Private.Uri/src/System/UriScheme.cs (hunk 10 above)
Comment (GrabYourPitchforks) [reply src/libraries/System.Private.Uri/src/System/UriScheme.cs]: I don't think so, as I believe these are intended to represent tokens / identifiers that shouldn't be treated in a linguistic-aware fashion. @scalablecory, would you be able to confirm?

Comment (lpereira) [reply src/libraries/System.Private.Uri/src/System/UriScheme.cs]: According to [RFC3986](https://tools.ietf.org/html/rfc3986), the scheme part of an URI has the following grammar:

      scheme      = ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )

Moreover, it says this about the kind of characters that are considered part of URIs:

                             A URI is a sequence of characters from a
      very limited set: the letters of the basic Latin alphabet, digits,
      and a few special characters. 

So I think it's safe to say that being culture invariant here is the way to go.

**Comment by GrabYourPitchforks** [comment]
@dotnet/jit-contrib Any thoughts as to Steve's comment at https://github.com/dotnet/runtime/pull/31968#discussion_r376682441? I can speak to impacting steady-state perf, but I don't really know how to reason about startup perf.

**Comment by stephentoub** [comment]
To clarify, my question is really whether `typeof(T).IsValueType` is going to be treated as an intrinsic and substituted for even in min opts, or whether by adding a bunch of these calls we're actually going to be doing more reflection and invoking IsValueType on a bunch more code paths.  If it's the former, great.

**Comment by GrabYourPitchforks** [comment]
A quick benchmark showing that `string.ToUpperInvariant` is able to take advantage of the optimized codegen of calling `TextInfo.Invariant` directly:

```cs
[Benchmark]
[Arguments("hello")]
[Arguments("XYZ")]
public string ToUpperInvariant(string s) => s.ToUpperInvariant();
```

```asm
; BEFORE - Benchmark.ToUpperInvariant(string)
0007ffd`0073a550 488bca          mov     rcx,rdx
00007ffd`0073a553 3909            cmp     dword ptr [rcx],ecx
00007ffd`0073a555 e9de9ed7ff      jmp     CLRStub[MethodDescPrestub]@7ffd004b4438 (00007ffd`004b4438) ; string.ToUpperInvariant (non-virtual dispatch)
; BEFORE - string.ToUpperInvariant()
00007ffd`00739dd0 56              push    rsi
00007ffd`00739dd1 4883ec20        sub     rsp,20h
00007ffd`00739dd5 488bf1          mov     rsi,rcx
00007ffd`00739dd8 48b98024ce1691020000 mov rcx,29116CE2480h
00007ffd`00739de2 488b09          mov     rcx,qword ptr [rcx]
00007ffd`00739de5 48b8c8c56100fd7f0000 mov rax,7FFD0061C5C8h
00007ffd`00739def ff10            call    qword ptr [rax] ; CultureInfo.get_TextInfo (non-virtual dispatch)
00007ffd`00739df1 488bc8          mov     rcx,rax
00007ffd`00739df4 488bd6          mov     rdx,rsi
00007ffd`00739df7 3909            cmp     dword ptr [rcx],ecx
00007ffd`00739df9 4883c420        add     rsp,20h
00007ffd`00739dfd 5e              pop     rsi
00007ffd`00739dfe e9fde6d8ff      jmp     CLRStub[MethodDescPrestub]@7ffd004c8500 (00007ffd`004c8500) ; TextInfo.ToUpper (non-virtual dispatch)

; AFTER - Benchmark.ToUpperInvariant(string)
00007ffd`0075a280 8b0a            mov     ecx,dword ptr [rdx]
00007ffd`0075a282 48b930267a1ff9010000 mov rcx,1F91F7A2630h ; string.ToUpperInvariant inlined into caller
00007ffd`0075a28c 488b09          mov     rcx,qword ptr [rcx]
00007ffd`0075a28f 3909            cmp     dword ptr [rcx],ecx
00007ffd`0075a291 e962e2d8ff      jmp     CLRStub[MethodDescPrestub]@7ffd004e84f8 (00007ffd`004e84f8) ; TextInfo.ToUpper (non-virtual dispatch)
```

|           Method |       Toolchain |     s |      Mean |     Error |    StdDev | Ratio | RatioSD |
|----------------- |---------------- |------ |----------:|----------:|----------:|------:|--------:|
| **ToUpperInvariant** |          **master** |   **XYZ** |  **7.081 ns** | **0.1886 ns** | **0.1764 ns** |  **1.00** |    **0.00** |
| ToUpperInvariant | toupper_tolower |   XYZ |  4.731 ns | 0.1721 ns | 0.1691 ns |  0.67 |    0.04 |
|                  |                 |       |           |           |           |       |         |
| **ToUpperInvariant** |          **master** | **hello** | **23.864 ns** | **0.3703 ns** | **0.3092 ns** |  **1.00** |    **0.00** |
| ToUpperInvariant | toupper_tolower | hello | 20.406 ns | 0.3830 ns | 0.3583 ns |  0.85 |    0.02 |

**Comment by GrabYourPitchforks** [comment]
@Jozkee @layomia Can you take a look at the commit https://github.com/dotnet/runtime/pull/31968/commits/f3e0b7c241d414fd977f28f66759c8d4c2d6a1bd, which is part of this PR? I want to confirm I'm not changing the intended meaning of the code.

**Comment by GrabYourPitchforks** [comment]
CI failures seem to be known issues.

**Comment by CarolEidt** [comment]
To answer (at in part) the startup question, the intrinsics are recognized in the importer, and AFAICT there are none that are only expanding when optimizing - @dotnet/jit-contrib are you aware of any issues with recognizing intrinsics in Tier0?

**Comment by AndyAyersMS** [comment]
> there are none that are only expanding when optimizing

Actually it's the other way round -- we generally won't expand intrinsics at Tier0/minopts:
```C++
    // Under debug and minopts, only expand what is required.
    if (!mustExpand && opts.OptimizationDisabled())
    {
        *pIntrinsicID = CORINFO_INTRINSIC_Illegal;
        return retNode;
    }
```
Must expand is only true for the recursive HW intrinsic methods and for a few of the old-style intrinsics.

**Comment by CarolEidt** [comment]
Thanks @AndyAyersMS - do you think that the lack of expansion for intrinsics in minopts/tier0 should inform their use? 

FWIW, I believe that all of the hw intrinsics are expanded, except those that are not directly supported (i.e. only expanded when `mustExpand` is true), as this code is executed before the part you cite above:
```
            if ((ni > NI_HW_INTRINSIC_START) && (ni < NI_HW_INTRINSIC_END))
            {
                GenTree* hwintrinsic = impHWIntrinsic(ni, clsHnd, method, sig, mustExpand);

                if (mustExpand && (hwintrinsic == nullptr))
                {
                    return impUnsupportedHWIntrinsic(CORINFO_HELP_THROW_NOT_IMPLEMENTED, method, sig, mustExpand);
                }

                return hwintrinsic;
```
The SIMD intrinsics seem to be recognized independent of optimization level also.

**Comment by AndyAyersMS** [comment]
The HW intrinsics are expanded in minopts/Tier0 only when they're seen as "recursive" methods; otherwise they're not expanded. So the instructions are not inlined into calling methods like they are when optimizing.

Other new-style intrinsics (like `Type.IsValueType` which I think is what prompted the questions above) don't behave this way and don't get recognized as intrinsics at minops/Tier0

> do you think that the lack of expansion for intrinsics in minopts/tier0 should inform their use?

Not quite sure what you are asking.

Arguably we should expand intrinsics at Tier0; it would probably lead to smaller and faster code and better jit throughput (see #9120). But then minopts and Tier0 would diverge...

**Comment by GrabYourPitchforks** [comment]
I don't believe `TextInfo.Invariant` is recognized as an intrinsic _per se_. Wouldn't this be going through the normal "evaluate static ctors and burn in the values of static readonly fields" mechanism, whatever that mechanism might be?

**Comment by AndyAyersMS** [comment]
If the class constructor have run before jitting, we'll burn static readonly values into codegen for numeric types, and optimize based on the actual type of the readonly statics for ref types (devirtualization, type tests, ...).

There is more we could do here for immutable or partially immutable types, eg static readonly array lengths, string lengths, string contents.

A static readonly ref type's fields are mutable, so we can't use them for optimization; likewise for static readonly structs or static readonly array elements.

**Comment by GrabYourPitchforks** [comment]
@AndyAyersMS Sure. I think I'm using incorrect terminology. In the assembly below, it's burning into the codegen not a reference to the `TextInfo.Invariant` object, but to the `TextInfo.Invariant` field (line 2 below), which still needs to be dereferenced (line 3 below) to get at the real object.

```cs
public string ToUpperInvariant(string s) => s.ToUpperInvariant();
```

```asm
00007ffd`0075a280 8b0a            mov     ecx,dword ptr [rdx]
00007ffd`0075a282 48b930267a1ff9010000 mov rcx,1F91F7A2630h
00007ffd`0075a28c 488b09          mov     rcx,qword ptr [rcx]
00007ffd`0075a28f 3909            cmp     dword ptr [rcx],ecx
00007ffd`0075a291 e962e2d8ff      jmp     CLRStub[MethodDescPrestub]@7ffd004e84f8 (00007ffd`004e84f8)
```

The concern AFAIK was whether introducing a static ctor on `TextInfo` would harm Tier 0 compilation. I don't quite know how to answer that question. The steady-state codegen looks quite optimal though, so I'm pleased with that at least. :)

**Comment by AndyAyersMS** [comment]
Addresses of statics are usually baked in to code -- for non-generic classes anyways.

Perhaps we should also look at the impact on R2R codegen? Prejitted code always has to check for class initialization.

I can run Tier0 and R2R diffs for you, but may not have results until tomorrow.

**Comment by GrabYourPitchforks** [comment]
> Perhaps we should also look at the impact on R2R codegen?

I'm open to whatever tests you think might be warranted. Right now I don't know how to run these tests (or even what tests to run!), but if you care to point me to them then I can take that under consideration for future changes.

**Comment by AndyAyersMS** [comment]
Diff summary [here](https://gist.github.com/AndyAyersMS/cde23e0b13e4d7240fde3cc2c1577b12). Diffs themselves are too large to share -- I can get you specific method diffs if you want. 

Overall impact on size is pretty minimal.

Tier0/Tier1 diffs for SPC sometimes include a few spurious diffs; I think that may be the case here.
